File: use-cases/vehicle-routing-pickup-delivery/src/main/java/org/acme/vehiclerouting/solver/justifications/VehicleCapacityJustification.java
package org.acme.vehiclerouting.solver.justifications;

import ai.timefold.solver.core.api.score.stream.ConstraintJustification;

public record VehicleCapacityJustification(String vehicleId, String visitId, int capacity, int demand,
                                           String description) implements ConstraintJustification {

    public VehicleCapacityJustification(String vehicleId, String visitId, int capacity, int demand) {
        this(vehicleId, visitId, capacity, demand, "Vehicle '%s' exceeded its max capacity by %s for visit '%s'."
                .formatted(vehicleId, demand - capacity, visitId));
    }
}

File: use-cases/vehicle-routing-kotlin/src/main/kotlin/org/acme/vehiclerouting/solver/justifications/VehicleCapacityJustification.kt
package org.acme.vehiclerouting.solver.justifications

import ai.timefold.solver.core.api.score.stream.ConstraintJustification

@JvmRecord
data class VehicleCapacityJustification(
    val vehicleId: String?, val capacity: Int, val demand: Int,
    val description: String
) : ConstraintJustification {
    constructor(vehicleId: String?, capacity: Int, demand: Int) : this(
        vehicleId, capacity, demand, "Vehicle '%s' exceeded its max capacity by %s."
            .formatted(vehicleId, demand - capacity)
    )
}